ompi/mca/coll/ftagree/coll_ftagree_earlyreturning.c

```diff
@@ -0,0 +1,3367 @@
+        }
+        if( nb > tree_size ) {
+            __tree_errors++;
+            fprintf(stderr, "TC %s -- %d/%d(%d): There is a cycle somewhere: I counted %d children, which is more than the tree size.\n",
+                    OMPI_NAME_PRINT(OMPI_PROC_MY_NAME), r, tree_size, tree[r].rank_in_comm, nb);
+            break;
+        }
+    }
+
+    if(display) {
+        fprintf(stderr, "\n");
+    }
+
+    return nb;
+}
+
+static void era_tree_check(era_tree_t *tree, int tree_size, int display)
+{
+    /** Tree consistency check */
+    int root = 0;
+    int children = 0;
+    int nodes = 0;
+    int r;
+
+    __tree_errors = 0;
+
+    for(r = 0; r < tree_size; r++) {
+        if( tree[r].rank_in_comm == -1 )
+            continue;
+        if( tree[r].parent == r ) {
+            root++;
+            if( tree[r].next_sibling != tree_size ) {
+                fprintf(stderr, "TC -- %d/%d: root cannot have a sibling (%d)\n",
+                        r, tree_size, tree[r].next_sibling);
+                __tree_errors++;
+            }
+        }
+        nodes++;
+        children += era_tree_check_node(tree, tree_size, r, display);
+    }
+    if( root != 1 ) {
+        fprintf(stderr, "TC -- There are %d roots in this tree. That's too many\n", root);
+        __tree_errors++;
+    }
+    if(children + 1 != nodes) {
+        fprintf(stderr, "TC -- There are %d nodes in this tree, and not %d\n", children+1, nodes);
+        __tree_errors++;
+    }
+
+    assert(__tree_errors == 0);
+}
+
+static char *era_debug_tree(era_tree_t *tree, int tree_size);
+#endif
+
+static void era_tree_fn_star(era_tree_t *tree, int tree_size)
+{
+    int m;
+    for(m = 0; m < tree_size; m++) { /**< O(nb_alive) comp. overhead */
+        tree[m].parent        = 0;
+        /** For tree_size == 1, 1 is also the terminating value, so this works */
+        tree[m].first_child   = m == 0 ? 1 : tree_size;
+        tree[m].next_sibling  = m == 0 ? tree_size : m + 1;
+    }
+}
+
+static void era_tree_fn_string(era_tree_t *tree, int tree_size)
+{
+    int m;
+    for(m = 0; m < tree_size; m++) { /**< O(nb_alive) comp. overhead */
+        tree[m].parent        = m > 0 ? m-1 : 0;
+        tree[m].first_child   = m + 1;
+        tree[m].next_sibling  = tree_size;
+    }
+}
+
+static void era_tree_fn_binary(era_tree_t *tree, int tree_size)
+{
+    int m;
+
+    for(m = 0; m < tree_size; m++) { /**< O(nb_alive) comp. overhead */
+        tree[m].parent        = m > 0 ? (m-1)/2 : 0;
+        tree[m].first_child   = (2*m+1 < tree_size) ? 2*m+1 : tree_size;
+        tree[m].next_sibling  = (m%2 != 0 && m+1 < tree_size ) ? m + 1 : tree_size;
+    }
+}
+
+typedef void (*era_tree_fn_t)(era_tree_t *, int);
+static era_tree_fn_t era_tree_fn = era_tree_fn_binary;
+
+typedef struct {
+    int rep;
+    int size;
+    era_tree_t *tree;
+} hierarch_tree_info_t;
+
+static void era_call_tree_fn(era_agreement_info_t *ci)
+{
+#if 0
```

Comment:
I would remove  the hierarchical tree implementation as it is commented out (and maybe keep it in a separate branch?)

Reply:
It stopped working after an update to ORTE that removed the locality information from the orte_process_info tables. In particular the `orte_ess.proc_get_daemon()` for non-self procs would not work. 

I believe a similar feature is now available in PMIX; @dong0321 do you remember? We want something  that can tell us if a process is colocated in the same host (i.e. the same test that would tell us we can use SM).

Reply:
This information is indeed available in PMIx. Ralph sent few emails about it on the ompi-devel mailing list (search for subject "Taking advantage of PMIx: Hierarchical collective support").